lib/lists.nix

```diff
@@ -116,6 +116,57 @@ rec {
   */
   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
 
+  /* Like `filter`, but with an index. O(n) complexity.
+
+     Type: filter :: (int -> a -> bool) -> [a] -> [a]
+
+     Example:
+       ifilter0 (i: v: i == 0 || v > 2) [ 1 2 3 ]
+       => [ 1 3 ]
+  */
+  ifilter0 =
+    ipred:
+    input:
+    let
+      # Whether each element in the input list is included
+      includeList = imap0 ipred input;
+
+      # The length of output, calculated by iterating over includeList and counting the true's
+      outputLength = foldl' (count: include:
+        if include
+        then count + 1
+        else count
+      ) 0 includeList;
```

Comment:
If this is indeed more efficient/otherwise preferrable to `length ∘ filter`, maybe we should make it a function as well

Reply:
What function are you thinking of here? Because this code here counts the number of `true` elements, I don't think that has a lot of applications

Reply:
```
      count = predicate: foldl' (count: el:
        if predicate el
        then count + 1
        else count
      ) 0;
```

Reply:
Oh yeah, I think that's worthwhile

Reply:
Turns out it [already exists](https://nixos.org/manual/nixpkgs/stable/#function-library-lib.lists.count)! Now using it instead :)